### PR TITLE
Single source hybrid property

### PR DESCRIPTION
**Pull request by CheatEx**

### CODE REVIEW

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 1]
```diff
@@ -177,53 +217,55 @@ public void run() {
...
-          mySource.add(i, myFrom.apply(i, p));
+          if (i == forceToSource) {
+            forceUpdated = true;
+          }
+          sourceAdd(i, p);
         } else {
-          // A new token added.
-          myTokens.add(p);
-          mySource.add(myFrom.apply(i, p));
+          throw new IllegalStateException("A printer generated more tokens than where is source.");
```
Comment (timzam) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: I would use this text:
"A printer generated more tokens than there are in the source."
or 
"A printer generated more tokens than the source."

**Comment by timzam** [comment]
Looks good to me.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+  private final Parser<? extends ModelT> myParser;
+  private final PrettyPrinter<? super ModelT> myPrinter;
+  private final ParsingContextFactory myParsingContextFactory;
+
+  private boolean myInSync = false;
+  private Listeners<EventHandler<? super PropertyChangeEvent<ModelT>>> myHandlers;
+
+  public PrettyHybridProperty(
+      ObservableList<SourceT> source,
+      final IndexedTransform<SourceT, Token> to,
```
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: In our frameworks the 'from' object usually goes before 'to' object.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+          }
+        });
+      }
+
+      @Override
+      public void onItemSet(final CollectionItemEvent<? extends SourceT> event) {
+        inSync(new Runnable() {
+          @Override
+          public void run() {
+            myTokens.set(event.getIndex(), to.apply(event.getIndex(), event.getNewItem()));
```
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: myRemoveHandler should be applied to event.getOldItem()

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+  private boolean myInSync = false;
+  private Listeners<EventHandler<? super PropertyChangeEvent<ModelT>>> myHandlers;
+
+  public PrettyHybridProperty(
+      ObservableList<SourceT> source,
+      final IndexedTransform<SourceT, Token> to,
+      IndexedTransform<Token, SourceT> from,
+      Handler<Integer> removeHandler,
+      Handler<Runnable> updateHandler,
```
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: what for is this handler?

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    myUpdateHandler.handle(new Runnable() {
+      @Override
+      public void run() {
+        mySource.remove(index);
+      }
+    });
+  }
+
+  public interface IndexedTransform <SourceT, TargetT> {
+    TargetT apply(int index, SourceT source);
```
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: is index required for implementation?

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+              updateFromPretty(index);
+            } else {
+              sourceRemove(index);
+            }
+          }
+        });
+      }
+    }
+
+    private void updateFromPretty(final int forceToSource) {
```
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: no need for final here

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+      PrettyPrinterContext<? super ModelT> printCtx = new PrettyPrinterContext<>(myPrinter);
+      printCtx.print(myValue);
+      List<Token> prettyTokens = printCtx.tokens();
+      boolean forceUpdated = false;
+      for (int i = 0; i < prettyTokens.size(); i++) {
+        final Token p = prettyTokens.get(i);
```
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: no need for final here

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+        if (i < myTokens.size() && i < mySource.size()) {
+          // Token exists in both source and tokens.
```
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: and ? tokens

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+          if (!Objects.equals(p, myTokens.get(i))) {
+            myTokens.set(i, p);
+            if (i == forceToSource) {
+              forceUpdated = true;
+            }
+            sourceSet(i, p);
+          }
+        } else if (i < myTokens.size() && i >= mySource.size()) {
+          // Token just appeared in tokens and doesn't exists in source.
```
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: 'doesn't exist'

**Comment by katepol** [comment]
When we are done with review fixes, I'd like you to make 1 commit for all the changes.

**Comment by katepol** [comment]
Are you going to remove ParsingHybridProperty after these changes are applied?

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 2 above)
Comment (azlobinjb) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Renamed to "read" and "write" to clarify the purpose.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 4 above)
Comment (azlobinjb) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Remove handler need to communicate token removal to application. Update handler needed to wrap source operation, for instance into OT transaction.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 5 above)
Comment (azlobinjb) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Yes.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 6 above)
Comment (azlobinjb) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Fixed.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 7 above)
Comment (azlobinjb) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Fixed.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 8 above)
Comment (azlobinjb) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Sorry, I don't understand the problem here.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 9 above)
Comment (azlobinjb) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Fixed.

**Comment by azlobinjb** [comment]
> When we are done with review fixes, I'd like you to make 1 commit for all the changes.
We will lost the entire discussion than, are you sure?

> Are you going to remove ParsingHybridProperty after these changes are applied?
I'm thinking about it. ParsingHybridProperty has simpler interface and more straightforward implementation. It still could be useful for some applications.

**Comment by timzam** [comment]
Did you fix the wording for the Illegal State exception message?

**Comment by CheatEx** [comment]
@timzam Yes.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 4 above)
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: OT transaction already handles all editor events. I don't see the reason.

**Comment by katepol** [comment]
We won't loose the discussion. It will be buried together with this pull request :)

**Comment by katepol** [comment]
Pretty- and Parsing- HybridProperties have code in common. May be a base class will help to show the difference between Pretty- and Parsing- HybridProperties functionality. What do you think?

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 5 above)
Comment (katepol) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Could you provide the example, please?

**Comment by CheatEx** [comment]
@katepol As I can see `get`/`addHandler`/`updateValue` methods could be extracted to a base class. For 2 descendants that wouldn't save any LOCs and will only harden reading.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 8 above)
Comment (CheatEx) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: No word missing - tokens is a field name.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 5 above)
Comment (CheatEx) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: If implementation allocates some resources for source than this index could be used as an id for maintenance and dispose.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java (hunk 4 above)
Comment (CheatEx) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java]: Some changes come here without enclosing transaction.

About index handling see https://github.com/JetBrains/jetpad-projectional/pull/230#discussion_r59554682

**Comment by katepol** [comment]
OK, then we need better names: both of these hybrid properties are parsing.

**Comment by timzam** [comment]
We shouldn't touch commits that are public. We should continue work in a pull request until it's accepted. We should close the pull request only if we decide to discard the approach completely.
It's a good idea to squash commits before they become public if they are close enough and represent more or less the same change.

**Comment by katepol** [comment]
Ok so the decision is to squash these commits when we all are agreed?

**Comment by CheatEx** [comment]
@timzam Problem with this PR is that I found 2 bugs after submit and that notes came in 3 waves. We need to choose what's more important for history - transparency or compactness.

**Comment by katepol** [comment]
Ok to me.

**Pull request closed without merging**